CommandExecutor improvements

akiraveliara opened the pull request:
- removes the dynamic invocation when executing a command
- abstracts CommandExecutor to pave the way for multiple future executors
- removes ReturnValue from CommandExecutedEvent, which was broken anyways (holding on to a ValueTask after it is awaited is a pretty terrible idea for multiple reasons)
- executes all pre-execution checks and reports an aggregate error, currently using placeholders until checks are reworked, instead of executing them one by one

i haven't properly tested this yet, awaiting review first, and i'd also love if you have a test bot/example bot that covers the scenarios you care about, as opposed to the scenarios i think you might care about

akiraveliara committed:
il emit, abstract command executor

akiraveliara committed:
improve invocation, pre-execution checks and fix a nullability error

OoLunar commented:
> removes ReturnValue from CommandExecutedEvent, which was broken anyways (holding on to a ValueTask after it is awaited is a pretty terrible idea for multiple reasons)

As discussed within the Discord, please try to reimplement this on a best-effort basis. If this becomes a task that wastes too much time, don't worry about it. We can always come back to it later.

> i haven't properly tested this yet, awaiting review first, and i'd also love if you have a test bot/example bot that covers the scenarios you care about, as opposed to the scenarios i think you might care about

I used the commands within the [example bot](https://github.com/OoLunar/DSharpPlus.CommandAll/tree/a2854ebd22af8a776a5c2c480e54548c2e26862d/examples/DSharpPlus.CommandAll.Basics) for quick testing. Ideally there would be a test suite with a custom `TestCommandContext` that immediately returned, however I'm not sure where that'd go within the mono repo. It's on the PR's todo list, needless to say.

OoLunar commented on DSharpPlus.Commands/ICommandExecutor.cs:
```diff
@@ -0,0 +1,23 @@
+    /// <param name="context">The context of the command.</param>
+    /// <param name="awaitCommandExecution">Whether to not return until the command has finished executing.</param>
+    /// <param name="cancellationToken">The cancellation token to use.</param>
+    public ValueTask ExecuteAsync
```
Could this be a delegate instead of an interface

akiraveliara replied:
it could, having an interface is probably preferable though since it gives DSP.C a concrete instance of a concrete type to work with

OoLunar replied:
Future proofing, got it.

OoLunar commented on DSharpPlus.Commands/CommandExecutor.cs:
```diff
@@ -54,37 +51,25 @@ private static async ValueTask WorkerAsync(CommandContext context)
             {
-                // Reverse foreach so we execute the top-most command's checks first.
-                for (int i = checks.Count - 1; i >= 0; i--)
+                if (!await checks[i].ExecuteCheckAsync(context))
```
User implemented checks are executed with the expectation to throw. With the removal of the try/catch statement, how will the behavior change?

akiraveliara replied:
that'll have to be updated once non-converter attributes are a thing

OoLunar commented on DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:
```diff
@@ -0,0 +1,142 @@
+            return GetTaskFunc(method);
+        }
+
+        throw new InvalidOperationException
```
Does this mean a method's return type *must*  be a `Task` or `ValueTask`? If so, the command builder's `From` method should be updated to check this.

akiraveliara replied:
i don't think they should, other executors are should be free to use other return types imo

OoLunar commented on DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:
```diff
@@ -0,0 +1,142 @@
+
+        throw new InvalidOperationException
+        (
+            $"This command executor only supports ValueTask and Task return types for commands, found " +
```
I am blind.

OoLunar commented on DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:
```diff
@@ -0,0 +1,142 @@
+            // not static implies declared within a type
+            if (method.DeclaringType!.IsValueType)
+            {
+                il.Emit(OpCodes.Unbox_Any, method.DeclaringType!);
```
If a delegate is passed to the command builder, the declaring type can be null

akiraveliara replied:
that... wasn't supposed to be what's pushed, did i accidentally revert an edit

akiraveliara replied:
oh no, we're fine, i didn't check the full context - the IsStatic check already makes sure that the declaring type is not null, since every instance method necessarily lives in a type

OoLunar replied:
Does this include anonymous delegates? I.E:

```cs

public Command BuildFakeCommand()
{
  CommandBuilder builder = new();
  builder.Name = "Test";

  // How the fuck do you explicitly declare the return type?
  builder.WithDelegate(async ValueTask(CommandContext context)
    => await context.RespondAsync("Ping!"));

  return builder.Build();
}

```

akiraveliara replied:
now it does!

OoLunar commented on DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:
```diff
@@ -0,0 +1,142 @@
+        // actually call the command method
+        il.Emit(OpCodes.Call, method);
+
+        il.Emit(OpCodes.Ret);
```
Returns the value of the called method or returns with nothing (`void`)?

akiraveliara replied:
returns the item on the execution stack, which is the return value of the `call`

OoLunar replied:
Wonderful.

OoLunar reviewed (approved):
Generally LGTM, just a few concerns here and there.

akiraveliara committed:
remove the exception throwing block

akiraveliara commented:
tested :)

akiraveliara committed:
support anon delegates

it's... not exactly fast, nor is it very nice, but it works unless one actively tries to break it. this code might also break if the compiler changes the way they're emitted, in which case we're in some trouble
